contracts/NFTMarketplace.sol

```diff
@@ -0,0 +1,173 @@
+        uint256 price;
+        bool fulfilled;
+        bool cancelled;
+    }
+
+    event OfferListed(
+        uint256 offerId,
+        uint256 id,
+        address user,
+        uint256 price,
+        bool fulfilled,
+        bool cancelled
+    );
+
+    event OfferFilled(uint256 offerId, uint256 id, address newOwner);
+    event OfferCancelled(uint256 offerId, uint256 id, address owner);
+    event ClaimFunds(address user, uint256 amount);
+
+    event CollectionCreated(uint256 indexed collectionId, string collectionName);
+
+    event MarketNftCreated(
+        uint256 indexed tokenId,
+        uint256 collectionId,
+        uint256 price,
+        bool forSale
+    );
+
+    constructor(address _marketItemAddress) {
+        marketItem = MarketItem(_marketItemAddress);
+        owner = payable(msg.sender);
+    }
+
+    modifier HasTransferApproval(uint256 _tokenId) {
+        require(marketItem.getApproved(_tokenId) == address(this), 'Market is not approved');
+        _;
+    }
+
+    modifier IsForSale(uint256 _tokenId) {
+        require(!nftLedger[_tokenId].forSale, 'Item is already sold');
+        _;
+    }
+
+    function createCollection(string memory _collectionName) external {
+        require(bytes(_collectionName).length != 0, 'Collection cannot be empty');
+
+        _collectionIds.increment();
+        uint256 newCollectionId = _collectionIds.current();
+
+        collectionLedger[newCollectionId] = Collection(_collectionName);
+
+        collectionIds.push(newCollectionId);
+        emit CollectionCreated(newCollectionId, _collectionName);
+    }
+
+    function createMarketItem(uint256 _collectionId, string memory _uri) external {
+        uint256 tokenId = marketItem.safeMint(msg.sender, _uri);
+
+        nftLedger[tokenId] = MarketNft(tokenId, _collectionId, 0, false);
+
+        nftLedgerIds.push(tokenId);
+        emit MarketNftCreated(tokenId, _collectionId, 0, false);
+    }
+
+    function setApproval(address _marketplaceContract, uint256 _tokenId) external {
+        marketItem.approve(_marketplaceContract, _tokenId);
+    }
+
+    function listItem(uint256 _tokenId, uint256 _price) private {
+        // - approval FE marketItem
+        nftLedger[_tokenId].forSale = true;
+        nftLedger[_tokenId].price = _price;
+    }
+
+    function buyItem(uint256 _tokenId)
+        external
+        payable
+        IsForSale(_tokenId)
+        HasTransferApproval(_tokenId)
+    {
+        uint256 price = nftLedger[_tokenId].price;
+
+        require(msg.value >= price, 'Not enough funds sent');
+
+        nftLedger[_tokenId].forSale = true;
+        marketItem.transferFrom(marketItem.ownerOf(_tokenId), msg.sender, _tokenId);
+
+        payable(msg.sender).transfer(msg.value);
+
+        marketFee[marketItem.ownerOf(_tokenId)] += LISTING_FEE;
+        credits[marketItem.ownerOf(_tokenId)] += msg.value;
+    }
+
+    function cancelItem(uint256 _tokenId) external {
+        nftLedger[_tokenId].forSale = false;
```

Comment:
Need to make some checks here:
* is the msg.sender owner of the NFT
* is the item actually for sale